lib/prelude.dx

```diff
@@ -964,6 +964,13 @@ def tile1 (fTile : (t:(Tile n l) -> {|eff} m=>l=>a))
 data List a =
   AsList n:Int elements:(Fin n => a)
 
+instance [Eq a] Eq (List a)
+  (==) = \(AsList nx xs) (AsList ny ys).
+    if nx /= ny
+      then False
```

Comment:
There is one interesting question when dealing with lists. What's the result of
```
(AsList -1 []) == (AsList -2 [])
```
Note that both definitions are well-typed, because both `Fin -1` and `Fin -2` represent empty index sets (but I don't think they compare equal in our type system?).

Reply:
Wow, I hadn't thought of that edge case!  The most satisfying solution that comes to my mind is to add a `Nat` type, and only let `Fin` take `Nat`s.  Or at least make the `List` constructor use a `Nat`.

Reply:
I think we should seriously consider adding a `Nat` type, not just for `Fin : Nat -> Type` but for ordinal indices too. It would let us get rid of some [clamping logic](https://github.com/google-research/dex-lang/blob/main/src/lib/Builder.hs#L845-L848) that has a real run-time cost whenever we use `fromOrdinal`. Instead, the clamping would only need to happen if you're coverting from `Int` to `Nat`, which might be fairly rare. Adam and I talked about it briefly some time ago and Adam had some reservations about surprising implications of having unsigned int types. Maybe we should open a discussion for it?